core/src/main/scala/org/zalando/grafter/Queries.scala

```diff
@@ -0,0 +1,50 @@
+package org.zalando.grafter
+
+import org.bitbucket.inkytonik.kiama.rewriting.MemoRewriter._
+
+import scala.collection.mutable.ListBuffer
+import scala.reflect.ClassTag
+
+trait Query {
+  
+  /** collect all distinct nodes (based on reference equality) of type T in a graph */
+  def collect[T : ClassTag, G](graph: G): List[T] = {
+    var seen: Vector[Int] = Vector.empty
+
+    // this stores the results
+    val results: ListBuffer[T] = new ListBuffer[T]
+
+    val collectStrategy =
+      everywheretd(strategy[Any] {
+        case t: T if Reflect.implements(t) =>
```

Comment:
Not really related to this issue, but I always stumble over how `implements` is used. It would be nice if instead we could write 

    def collect[T, G](graph: G)(implicit classTag: ClassTag[T]): List[T] = {
        ...
        case t: T if t implements classTag => ...
    }